### PR TITLE
Add forc `BuildPlan`. Change GitHub dependencies to general git support. Add `Forc.lock` for more deterministic builds.

### LINKED ISSUE #746: Namespace dependencies directory
Currently, dependencies are placed in `~/.forc/<dependency ID>`. This should be namespaced into something like `~/.forc/git/checkouts/` (to mirror what Cargo does, but could be something different). Namespacing will allow other things to be placed in the Forc directory without collision.

### LINKED ISSUE #798: `forc build` - Create "build plan" before beginning compilation
Currently, the process of compiling dependencies is intertwined with fetching the program's immediate child dependencies before then resolving and compiling child dependencies recursively.

It could be worth splitting this step into:

1. Construct "build plan" as a graph of all dependencies.
2. Compile the graph of dependencies in topological order.

By creating a build plan as a separate first step there are a few potential benefits:

- Detect cycles before beginning compilation.
- Support dependency graph visualisations (e.g. `.dot` files) even if one or more dependencies don't compile successfully. 
- More easily avoid compiling duplicate dependencies more than once.
- Better prepare for potentially supporting something like `Forc.lock` and reproducible builds in the future.
- Pave the way for compiling multiple dependencies concurrently.
- Friendlier interop with hermetic build systems like Nix.

### LINKED ISSUE #799: Allow for "pinning" dependencies to better support reproducible builds, i.e. `Forc.lock`
Similar to cargo's `Cargo.lock`.

*Depends on #798.*

### LINKED ISSUE #829: Change forc's `git` dependencies to not rely on the GitHub REST API. Use git tags to pin to versions instead.
Currently `forc` doesn't really support `git` dependencies generally, but rather calls into the GitHub REST API.

As a part of https://github.com/FuelLabs/sway/pull/825, I'm aiming to switch us to a more genuine/generalised git support like cargo has, however doing so will mean changing the way we pin to specific versions for git dependencies.

The current way that `std` and `core` git dependencies are pinned relies on using the GitHub-specific REST API to fetch the pinned "release" tarball. Under cargo git dependencies (and git itself), the term "version" doesn't really mean anything, as git dependencies are normally pinned to a particular commit either directly (`rev`), under a `tag` or under a given git `branch`. In order to emulate the current behaviour where we pin `std` and `core` to a particular `version`, I'd recommend we instead add a git tag to the commit that publishes the new version and pin the dependency using the version tag (e.g. `tag = "v0.1.0"`).

### PR DESCRIPTION
**Pull request by mitchmindtree**
OK, this should be ready for review!

Apologies for the massive PR! Originally I only intended to refactor `forc_build` so that the process of fetching dependencies and determining the package graph could be distinct from the process of compilation (see #798), however upon beginning I figured it made sense to also address #829 (remove GitHub-only git repo requirement) as the process of building the package graph is tightly coupled with the fetching of those packages. This lead to addressing #746 (adding proper namespace for forc's git checkouts) as the existing directory structure didn't make sense for more generalised git support. In order to allow for safely caching/re-using pre-checked-out builds, I realised I needed the ability to pin specific commits, so I also ended up addressing #799 as a result. In hindsight, I would have been better off splitting off #829 into a future PR :sweat_smile: but it's done now.

*Edit: Keep in mind about ~2K of the additions are just the new `Forc.lock` files.*

---

## `BuildPlan` (#798)

The new `BuildPlan` type is an internal detail that represents a forc project's entire build plan. This includes the package graph, a map from all packages to a local copy of their src, and the compilation order determined by performing a toposort of the package graph.

### The package `Graph`

The package graph represents the dependencies between all packages within the project. Each node represents a *pinned* package. The edge *a -> b* implies that *a* depends on *b*.

## Pinning, `Lock` & `Forc.lock` (#799)

Pinning a dependency means fetching its source, determining the latest commit for the given branch (or the latest version that adheres to semver in the case of upcoming registry support) and then ensuring that we can deterministically make use of this same commit or version within future builds.

To achieve this, this PR introduces the `Forc.lock` file (directly inspired by `Cargo.lock`) along with a `Lock` structured representation.  The `Lock` type is a toml-serialization-friendly version of the package graph that, when serialized to TOML and written to `Forc.lock`, acts as a human-readable, git-diff-friendly representation of the project's graph of pinned dependencies.

By using a lock file to represent pinned dependencies, we can allow for better reproducibility in forc projects without requiring that users write the exact pinned version or commit hashes manually. Instead, users can run `forc update` to update the commit (or version) for each pinned dependency automatically.

`forc build` will automatically create a new `Forc.lock` file if one didn't exist, or if the existing file doesn't match the current set of dependencies in the `Forc.toml` manifest. Both `forc update` and `forc build` will print a diff of the `Lock` when updating the `Forc.lock` file. This involves first printing removed packages in red, and added packages in green.

## GitHub -> git (#829, #746)

This PR removes all the GitHub-specific code under handling of git dependencies in favour of more general git support using the `git2` crate.

Previously, forc used the `version` field for git dependencies to imply a GitHub "release". The GitHub REST API was used to fetch this release. This PR removes this behaviour, as `version` doesn't really have any particular meaning under git itself. Instead, this adds support for pinned git branches *and* git **tag**s.

Unfortunately, `git2` does depend on openssl cc @adlerjohn. It looks like there's some recent interest in landing a rustls+hyper backend discussed [here](https://github.com/rust-lang/git2-rs/issues/623) along with a WIP that could potentially resolve this dependency in the future.

## E2E tests & examples

For now, examples and E2E tests have been updated to use git tags instead. However, following this PR we should consider updating all examples and E2E tests to point to master and instead commit the Forc.lock file to pin each project's versions.

One issue that updating the tests and examples has revealed is that currently, the `std` dependency doesn't actually specify a pinned version (or tag) for `core`. As a result, all E2E tests and examples that use some version or tag instead of the `master` branch for the `core` dependency end up with two versions of `core` in their `Forc.lock` file: 1. the `core` master branch as defaulted to in the `std` library's dependencies and 2. the `core` version listed under the top-level project's dependencies. Fortunately this doesn't seem to cause any issues in most examples. Those few that did have issues have been updated.

Closes #746.
Closes #798.
Closes #799.
Closes #829.

---

# TODO

- [x] See if we can pass the new graph by immutable reference throughout compilation process and remove old mutable hashmap reference.
- [x] Investigate whether or not `Namespace` population is affected by this change.
- [x] Remove old replaced dependency compilation code.
- [x] Handle offline mode (error on remote deps for now).
- [x] Move package-fetching logic into a separate module.
- [x] Work out why `if_elseif_else` test example isn't building. Edit: This is due to the `version` field in `git` dependencies no longer having meaning. See #829 and #831.
- [x] Remove old github-specific depedency handling in favour of the new generalised `git` handling approach. See #829.
- [x] Write the graph of pinned packages to a `Forc.lock` file inspired by `Cargo.lock` #799. Always attempt to load the graph from the lock file first. On `forc update`, update the lock file. The lock file feature must be implemented in order to allow for re-using pre-fetched dependencies - otherwise we need to re-fetch dependencies on every build which hugely increases build times, especially for E2E tests.
- [x] Make `forc update` simply update the `Forc.lock` file.
- [x] Update workspace lock file for changes to `forc`'s `[dependencies]`.
- [x] Check that building with a `Forc.lock` works even if `~/.forc/git/checkouts` doesn’t exist.
- [x] Commit `Forc.lock` files for all examples and E2E tests for better reproducibility and faster compilation by avoiding fetching the same deps for every package.

### Follow-up

- [ ] Open issue for `forc build --plan-only` flag or similar that outputs the `pkg_graph` via JSON or DOT. For now, downstream can always reconstruct the graph from the `Forc.lock`, but it's not particularly convenient.
- [ ] Add ability to update only specific packages throughout graph with `forc update --package <name>`.

### CODE REVIEW

File: forc/src/pkg.rs
Comment (mitchmindtree) [inline forc/src/pkg.rs:25]: Remove this comment.
```suggestion

```

File: forc/src/lock.rs
Comment (mitchmindtree) [inline forc/src/lock.rs:101]:
```suggestion
    /// Given a `Lock` loaded from a `Forc.lock` file, produce the graph of pinned dependencies.
```

File: forc/src/lock.rs
Comment (mitchmindtree) [inline forc/src/lock.rs:91]:
```suggestion
    /// Given a graph of pinned packages, create a `Lock` representing the `Forc.lock` file structure.
```

File: forc/src/pkg.rs
Comment (mitchmindtree) [inline forc/src/pkg.rs:297]: Consider not building the `PathMap` while constructing the `Graph`, and instead construct it afterwards using the new `graph_to_path_map` function above.

File: forc/src/pkg.rs
Comment (mitchmindtree) [inline forc/src/pkg.rs:658]:
```suggestion
// TODO: Update this to match behaviour described in `compile` doc comment above.
```

File: forc/src/pkg.rs
Comment (mitchmindtree) [reply forc/src/pkg.rs:334]: It makes sense to collect `PathMap` as we go as it doubles as a cache for already visited dependencies.

**Review by JoshuaBatty — DISMISSED** [review]
Epic work man! LGTM

**Comment by mitchmindtree** [comment]
Cheers chap! I might wait for a second review from @adlerjohn / @sezna / @digorithm before landing as this is quite a biggy.

**Comment by mitchmindtree** [comment]
OK, just rebased onto master again now that #860 has landed.

**Review by JoshuaBatty — APPROVED** [review]
Epic!! LGTM but i'll let some others chime in as it's a pretty big one.

**Review by sezna — APPROVED** [review]

**Pull request merged**
